### PR TITLE
RSDK-797 Add base wrapper

### PR DESCRIPTION
**Pull request by zaporter-work**
Ticket: https://viam.atlassian.net/browse/RSDK-797

Important decisions: 
1) For now I replaced common.v1.Vector3 with `typedef std::array<double, 3> Vector3` in `common/linear_algebra.hpp`. I am dubious about this for two reasons:
    - If we expose an Vector3, it should be a useful one. I am leaning towards the `boost::numeric::ublas` types
    - If we are following the proto file hierarchy, this should be defined in `common/common.hpp` however, that file doesn't exist, and we have instead chosen to create files like `common/utils.hpp` and `common/proto_type.hpp`
2) For testing, I have done a peeking-based testing architecture rather than the current behavior-based mocks. That was done because `base` has no good getter methods to test functionality. 
3) This does not have an example because I don't think there is any interesting behavior that is not already captured by the existing examples. Happy to add one if others think it would be useful.

EDIT: I have decided to move forward with the `boost::qvm::vec` trait applied to our custom simple Vector3 class.

Pre-merge:
Bring in refactoring changes. Right now there is so much changing that it would be a waste of time to try and keep up before it is stable

### CODE REVIEW

**Comment by zaporter-work** [comment]
I think this should be merged after Ethan's private rename refactor https://github.com/viamrobotics/viam-cpp-sdk/pull/89

File: src/viam/sdk/components/base/base.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+   public:
+    // functions shared across all components
+    static std::shared_ptr<ResourceSubtype> resource_subtype();
+    static Subtype subtype();
+
+    /// @brief Move a robot's base in a straight line by a given distance. This method blocks
+    /// until completed or cancelled
+    /// @param distance_mm Desired travel distance in millimeters
+    /// @param mm_per_sec Desired travel velocity in millimeters/second
+    virtual void move_straight(int64_t distance_mm, double mm_per_sec) = 0;
```
Comment (zaporter-work) [inline src/viam/sdk/components/base/base.hpp:51]: [q] Right now we do not use `boost::units`. There was an open question about that in the Servo PR. I have left it like this for now but I am happy to swap if that is the more modern approach.

File: src/viam/sdk/components/base/base.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    /// @brief Spins a robot's base by an given angle, expressed in degrees. This method blocks
+    /// until completed or cancelled
+    /// @param angle_deg Desired angle
+    /// @param degs_per_sec Desired angular velocity
+    virtual void spin(double angle_deg, double degs_per_sec) = 0;
```
Comment (zaporter-work) [inline src/viam/sdk/components/base/base.hpp:57]: Same `boost::units` applies here.

File: src/viam/sdk/components/base/base.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /// @brief Sets the linear and angular power of a base -1 -> 1 in terms of power for
+    /// each direction
+    /// @param linear Desired linear power percentage (-1 <= % <= 1) for each direction
+    /// @param angular Desired angular power percentage (-1 <= % <= 1) for each direction
+    virtual void set_power(Vector3 linear, Vector3 angular) = 0;
+
+    /// @brief Set the linear and angular velocity of a base
+    /// @param linear Desired linear velocity in mm per second for each direction
+    /// @param angular Desired angular velocity in degrees per second for each direction
+    virtual void set_velocity(Vector3 linear, Vector3 angular) = 0;
```
Comment (zaporter-work) [inline src/viam/sdk/components/base/base.hpp:68]: Here `boost::units` are complicated because we would want a vector of velocity units which is a very complex type to create.

File: src/viam/sdk/common/linear_algebra.hpp [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include <boost/numeric/ublas/storage.hpp>
+namespace viam {
+namespace sdk {
+
+/* typedef boost::numeric::ublas::bounded_array<double, 3> Vector3; */
+typedef std::array<double, 3> Vector3;
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/linear_algebra.hpp:7]: It's not immediately clear to me what the advantage of using the `boost` type is. Can you go into some more detail about what you think the `boost` type gives us that the `std::array` doesn't?

[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp:7]: In my mind, the benefit of the `boost` type is that it has additional member methods that will help both us and users. If we end up needing to do vector operations, it will be much nicer to have a real vector rather than a wrapped array. 

(This also applies to users. It is very likely that they will want to do things like add / cross-multiply vectors and it is more convenient if they do not have to convert into and out of a `std::array`)

[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/linear_algebra.hpp:7]: Makes sense to me, given the above I'm happy to switch to the `boost` type.

File: src/viam/sdk/components/base/base.hpp (hunk 1 above)
Comment (stuqdog) [reply src/viam/sdk/components/base/base.hpp:51]: I don't have any immediate thoughts about this. Will do a bit of research/thinking, but in the meantime I defer to @acmorrow

File: src/viam/sdk/components/base/server.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    void register_server(std::shared_ptr<Server> server) override;
+
+    std::shared_ptr<SubtypeService> get_sub_svc();
+
+    BaseServer() : sub_svc(std::make_shared<SubtypeService>()){};
+    BaseServer(std::shared_ptr<SubtypeService> sub_svc) : sub_svc(sub_svc){};
+
+   private:
+    std::shared_ptr<SubtypeService> sub_svc;
```
Comment (stuqdog) [inline src/viam/sdk/components/base/server.hpp:58]: Given recent refactors, I think we should remove `sub_svc` as a data member and `get_sub_svc()` as a method distinct from the one in `ResourceServerBase`.

**Review by stuqdog — CHANGES_REQUESTED** [review]
This looks good! A couple small comments but overall 👍

File: src/viam/sdk/components/base/server.hpp (hunk 5 above)
Comment (zaporter-work) [reply src/viam/sdk/components/base/server.hpp:58]: Agreed, was waiting for that refactor to go through. Related to this: https://github.com/viamrobotics/viam-cpp-sdk/pull/89#discussion_r1180673134

File: src/viam/sdk/components/base/client.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    request.mutable_angular()->set_y(angular.at(1));
+    request.mutable_angular()->set_z(angular.at(2));
+
+    grpc::Status status = stub_->SetPower(&ctx, request, &response);
+    if (!status.ok()) {
+        throw std::runtime_error(status.error_message());
+    }
+}
+
+void BaseClient::set_velocity(std::array<double, 3> linear, std::array<double, 3> angular) {
```
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/components/base/client.cpp:73]: Good catch. Will fix.

File: src/viam/sdk/common/linear_algebra.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp]: After more investigation, I have decided to use the `boost::qvm::vec` class. It has turned out quite well.

[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp]: Agreed. I have just better Vector3 class.

**Comment by acmorrow** [comment]
I will try to take a look at this early next week. Meanwhile, I'm happy with the pivot away from ublas because I have a need for tensor classes in my ml model work and I've opted explicitly *not* to use ublas and to use xtensor instead. So I'd prefer we not end up with both in the project. The QVM library isn't one I'm familiar with, but it seems at first approximation to be a tighter match with what this PR is undertaking anyway. I'll follow up on the `boost::units` thing in this PR too.

File: src/viam/sdk/common/linear_algebra.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#include <viam/sdk/common/linear_algebra.hpp>
+
+#include <viam/api/common/v1/common.pb.h>
+
+namespace viam {
+namespace sdk {
+
+double Vector3::x() const {
```
Comment (acmorrow) [inline src/viam/sdk/common/linear_algebra.cpp:8]: Per discussion, these trivial accessors probably do belong as inline. However, let's just add a TODO for that for now because I don't want to muddy the waters about what should/shouldn't be inline at this point. We can always un-inline them later.

File: src/viam/sdk/components/base/base.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    return Subtype(RDK, COMPONENT, "base");
+}
+
+namespace {
+bool init() {
+    Registry::register_subtype(Base::subtype(), Base::resource_subtype());
+    return true;
+};
+
+bool inited = init();
```
Comment (acmorrow) [inline src/viam/sdk/components/base/base.cpp:47]: `const bool`

File: src/viam/sdk/components/base/base.hpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    /// until completed or cancelled
+    /// @param angle_deg Desired angle
+    /// @param degs_per_sec Desired angular velocity
+    virtual void spin(double angle_deg, double degs_per_sec) = 0;
+
+    /// @brief Sets the linear and angular power of a base -1 -> 1 in terms of power for
+    /// each direction
+    /// @param linear Desired linear power percentage (-1 <= % <= 1) for each direction
+    /// @param angular Desired angular power percentage (-1 <= % <= 1) for each direction
+    virtual void set_power(Vector3 linear, Vector3 angular) = 0;
```
Comment (acmorrow) [inline src/viam/sdk/components/base/base.hpp:63]: Unless you are consuming these vectors, they should probably be passed by `const &`.

File: src/viam/sdk/components/base/client.hpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+   public:
+    void move_straight(int64_t distance_mm, double mm_per_sec) override;
+    void spin(double angle_deg, double degs_per_sec) override;
+    void set_power(Vector3 linear, Vector3 angular) override;
+    void set_velocity(Vector3 linear, Vector3 angular) override;
+    grpc::StatusCode stop() override;
+    grpc::StatusCode stop(AttributeMap extra) override;
+    bool is_moving() override;
+    AttributeMap do_command(AttributeMap command) override;
+    BaseClient(std::string name, std::shared_ptr<grpc::Channel> channel)
```
Comment (acmorrow) [inline src/viam/sdk/components/base/client.hpp:32]: ctor decl to the top, definition to .cpp, here and elsewhere.

File: src/viam/sdk/tests/test_base.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+//        \/
+// -- ComponentClient (Real)
+//
+// This is as close to a real setup as we can get
+// without starting another process
+//
+// The passed in lambda function has access to the ComponentClient
+//
+template <typename Lambda>
+void server_to_mock_pipeline(Lambda&& func) {
```
[the remark this answers is not in the document]
Comment (acmorrow) [reply src/viam/sdk/tests/test_base.cpp:55]: I know :)

Here, BTW, is what my version of it looks like in the ml model stuff:

```
template <typename F>
void client_server_test(std::shared_ptr<MockMLModelService> mock, F&& f) {
    const auto k_service_name = "mock_mlmodel_service";

    MLModelServiceServer mlmodel_server;
    mlmodel_server.resource_manager()->add(k_service_name, std::move(mock));

    grpc::ServerBuilder builder;
    builder.RegisterService(&mlmodel_server);
    auto server = builder.BuildAndStart();

    MLModelServiceClient client(k_service_name, server->InProcessChannel({}));
    std::forward<F>(f)(client);

    server->Shutdown();
}
```

There are three types in play here: `MockMLModelService`, `MLModelServiceServer`, and `MLModelServiceClient`.

If the mocks had a canonical way of naming the associated client and server types, then this could be templated on the mock type:

```
template <typename Mock, typename F>
void client_server_test(std::shared_ptr<Mock> mock, F&& f) {
    const auto k_service_name = "mocky mock";

    Mock::server_type server;
    server.resource_manager()->add(k_service_name, std::move(mock));

    grpc::ServerBuilder builder;
    builder.RegisterService(&server);
    auto server = builder.BuildAndStart();

    Mock::client_type client(k_service_name, server->InProcessChannel({}));
    std::forward<F>(f)(client);

    server->Shutdown();
}
```

Then you would have:

```
class FooMock {
public:
    using client_type = FooClient;
    using server_type = FooServer;
};
```

That protocol would let you lift this function to a test utility header.

I expect there are further typedefs along these lines that we will want, to make it possible to get easily from a Server type to its Client type and vice versa, and from the Client type to the base type. But I haven't fully formalized how I want that to work. I'm hoping to explore it further in the ml model tests I'm writing.

**Review by acmorrow — CHANGES_REQUESTED** [review]

File: src/viam/sdk/common/linear_algebra.hpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    Vector3(double x, double y, double z) : data_{x, y, z} {};
+    Vector3() : Vector3(0.0, 0.0, 0.0){};
+
+    double x() const;
+    double y() const;
+    double z() const;
+    void set_x(double x);
+    void set_y(double y);
+    void set_z(double z);
+    std::array<double, 3> data() const;
```
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp:20]: Oops -- good catch.

File: src/viam/sdk/common/linear_algebra.hpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    std::array<double, 3>& mutable_data();
```
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp:21]: From a language perspective, that is super cool. 

I talked with @stuqdog about this and right now I think the desired style is towards explicit mutability but I will keep this in mind going forward.

File: src/viam/sdk/common/linear_algebra.hpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+class Vector3 {
+   public:
+    Vector3(double x, double y, double z) : data_{x, y, z} {};
+    Vector3() : Vector3(0.0, 0.0, 0.0){};
+
+    double x() const;
+    double y() const;
+    double z() const;
+    void set_x(double x);
```
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp:17]: Similar to topic above, discussed with @stuqdog and decided to keep explicit setters. However, I added the ref return to allow for chaining of `set_` operators.

File: src/viam/sdk/common/linear_algebra.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (acmorrow) [reply src/viam/sdk/common/linear_algebra.hpp]: Hmm. OK. I'd say that is somewhat non-idiomatic for C++. Overloading accessors on `const` is pretty much the way it is done. It has the advantage that I don't need to know whether the object I'm operating on is `const` or not to call `data`. I get the right behavior and the right return type either way.

For instance, `std::vector` overloads `begin` such that when called on a non-const `std::vector` it returns a `std::vector::iterator` and when called on a `const` std::vector it returns a `std::vector::const_iterator`: https://en.cppreference.com/w/cpp/container/vector/begin. Don't be misled by the `cbegin` stuff that you will see there: that is there as a convenience shortcut to explicitly get a `const_iterator` when you may not have a const object, saving you a `const_cast` to `const`.

Basically, you want `const` to "flow": const objects return const references which in turn allows calling const methods. Making the name different means there is a point of divergence where you need to know the constness to know which method to call.

But I'll leave it to Ethan to make the final call.

File: src/viam/sdk/tests/test_base.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/tests/test_base.cpp:55]: Glad to hear! I'll keep copy/pasting for now but looking forward to simplifying in the future.

File: src/viam/sdk/components/base/base.hpp [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Base base.hpp "components/base/base.hpp"
+/// @brief A `Base` is the platform that the other parts of a mobile robot attach to.
+/// @ingroup Base
+///
+/// This acts as an abstract parent class to be inherited from by any drivers representing
+/// specific base implementations. This class cannot be used on its own.
+class Base : public ComponentBase {
```
Comment (acmorrow) [inline src/viam/sdk/components/base/base.hpp:41]: @stuqdog - A naming thing. This line got my attention, because `Base` is overloaded here. When used as part of the name `ComponentBase` it being used to mean base class, I believe. But `Base` is a kind of component too, and generally a term of art in the Viam world. It feels muddled.

How would you feel about (in another review) removing `Base` from `ResourceServerBase`, `ComponentBase`, and `ResourceBase`, which I think are the only places it is currently used?

File: src/viam/sdk/common/linear_algebra.hpp [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+namespace sdk {
+
+// In the future, we may wish to inline this whole class
+// for performance reasons.
+
+class Vector3 {
+   public:
+    using scalar_type = double;
+
+    Vector3(scalar_type x, scalar_type y, scalar_type z) : data_{x, y, z} {};
```
Comment (acmorrow) [inline src/viam/sdk/common/linear_algebra.hpp:16]: If this class is all out-of-line for now then these constructors probably should be too.

File: src/viam/sdk/common/linear_algebra.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (acmorrow) [reply src/viam/sdk/common/linear_algebra.hpp]: @stuqdog - I would like to hear your thoughts on my comments above.

**Review by acmorrow — APPROVED** [review]
LGTM

File: src/viam/sdk/components/base/base.hpp (hunk 15 above)
Comment (stuqdog) [reply src/viam/sdk/components/base/base.hpp:41]: That sounds reasonable to me, ticket [here](https://viam.atlassian.net/browse/RSDK-2971).

File: src/viam/sdk/tests/mocks/mock_base.hpp [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    void move_straight(int64_t distance_mm, double mm_per_sec) override;
+    void spin(double angle_deg, double degs_per_sec) override;
+    void set_power(const viam::sdk::Vector3& linear, const viam::sdk::Vector3& angular) override;
+    void set_velocity(const viam::sdk::Vector3& linear, const viam::sdk::Vector3& angular) override;
+    grpc::StatusCode stop() override;
+    bool is_moving() override;
+    viam::sdk::AttributeMap do_command(viam::sdk::AttributeMap command) override;
+    static std::shared_ptr<MockBase> get_mock_base();
+
+    // This variables allow the testing infra to `peek` into the mock
```
Comment (stuqdog) [inline src/viam/sdk/tests/mocks/mock_base.hpp:27]: (nit) s/This/These

File: src/viam/sdk/common/linear_algebra.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/linear_algebra.hpp]: @acmorrow your explanation makes sense and I think I'm leaning toward being convinced that we should overload on `const`ness. My concern is that this will lead to cases where people get mutable objects when they don't need/want mutability. Even if one has a non-`const` `T`, I could imagine they'd still want `const` behavior from certain calls. They could cast, but this requires extra code and greater awareness of the `const`ness of values. 

Given your statement that overloading on `const` is the more idiomatic approach, perhaps I am just speaking from a place if ignorance or trying to force C++ into something it isn't by using non-idiomatic paradigms from different languages. I'm certainly willing to be convinced here!

[the remark this answers is not in the document]
Comment (zaporter-work) [reply src/viam/sdk/common/linear_algebra.hpp]: Moving forward with `const`-based overload of `data()`

[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/linear_algebra.hpp]: Just spoke with @acmorrow offline, I have been convinced that overloading on `const` is the right approach here.

**Review by stuqdog — APPROVED** [review]
LGTM
